buildSrc/src/main/kotlin/embeddable.kt

```diff
@@ -94,6 +94,27 @@ fun Project.compilerDummyForDependenciesRewriting(
         body()
     }
 
+fun Project.unshadedCompiler(
```

Comment:
I was unable to get this to work using that example. The class files are not added to the jar and I'm not sure why

Reply:
```
plugins {
    java
}

val compilerModules: Array<String> by rootProject.extra

dependencies {
    compilerModules.forEach {
        embedded(project(it)) { isTransitive = false }
    }

    embedded(projectTests(":compiler:tests-common-jvm6")) { isTransitive = false }
    embedded(projectTests(":compiler:test-infrastructure")) { isTransitive = false }
    embedded(projectTests(":compiler:test-infrastructure-utils")) { isTransitive = false }
    embedded(projectTests(":compiler:tests-compiler-utils")) { isTransitive = false }
    embedded(projectTests(":compiler:tests-common-new")) { isTransitive = false }
}

publish()
runtimeJar()
sourcesJar()
javadocJar()
```

This script will include all compiler modules in the published jar + additional test framework modules specified bellow `compilerModules.forEach`

Reply:
This breaks because I cannot access the protobuf dependencies. I tried to fix this by adding a dependency on the protobuf module but it looks like the protobuf module is not a module at all (unlike the jvm for minimal testing) but published to the internal repository. I'm having a hard time locating how this dependency is specified in the serialization library in order to fix that dependency

Reply:
Have you tried to add `embedded(protobufFull())`? We do it like this when building compiler. You might also need to add `embedded(kotlinBuiltins())` but I'm not sure

Reply:
The `embedded(protobufFull())` solved the problem of missing protobuf dependencies but now it fails on resolving the LazyTopDownAnalyzer because of the DeclarationScopeProvider cannot be resolved (```Dependencies for `public org.jetbrains.kotlin.resolve.LazyTopDownAnalyzer(org.jetbrains.kotlin.resolve.BindingTrace,org.jetbrains.kotlin.resolve.DeclarationResolver,org.jetbrains.kotlin.resolve.OverrideResolver,org.jetbrains.kotlin.resolve.OverloadResolver,org.jetbrains.kotlin.resolve.VarianceChecker,org.jetbrains.kotlin.descriptors.ModuleDescriptor,org.jetbrains.kotlin.resolve.lazy.LazyDeclarationResolver,org.jetbrains.kotlin.resolve.BodyResolver,org.jetbrains.kotlin.resolve.lazy.TopLevelDescriptorProvider,org.jetbrains.kotlin.resolve.lazy.FileScopeProvider,org.jetbrains.kotlin.resolve.lazy.DeclarationScopeProvider,org.jetbrains.kotlin.resolve.QualifiedExpressionResolver,org.jetbrains.kotlin.resolve.IdentifierChecker,org.jetbrains.kotlin.config.LanguageVersionSettings,org.jetbrains.kotlin.resolve.deprecation.DeprecationResolver,java.lang.Iterable,org.jetbrains.kotlin.resolve.FilePreprocessor)` cannot be satisfied:
  [interface org.jetbrains.kotlin.resolve.lazy.DeclarationScopeProvider]
org.jetbrains.kotlin.container.UnresolvedDependenciesException: Dependencies for `public org.jetbrains.kotlin.resolve.LazyTopDownAnalyzer(org.jetbrains.kotlin.resolve.BindingTrace,org.jetbrains.kotlin.resolve.DeclarationResolver,org.jetbrains.kotlin.resolve.OverrideResolver,org.jetbrains.kotlin.resolve.OverloadResolver,org.jetbrains.kotlin.resolve.VarianceChecker,org.jetbrains.kotlin.descriptors.ModuleDescriptor,org.jetbrains.kotlin.resolve.lazy.LazyDeclarationResolver,org.jetbrains.kotlin.resolve.BodyResolver,org.jetbrains.kotlin.resolve.lazy.TopLevelDescriptorProvider,org.jetbrains.kotlin.resolve.lazy.FileScopeProvider,org.jetbrains.kotlin.resolve.lazy.DeclarationScopeProvider,org.jetbrains.kotlin.resolve.QualifiedExpressionResolver,org.jetbrains.kotlin.resolve.IdentifierChecker,org.jetbrains.kotlin.config.LanguageVersionSettings,org.jetbrains.kotlin.resolve.deprecation.DeprecationResolver,java.lang.Iterable,org.jetbrains.kotlin.resolve.FilePreprocessor)` cannot be satisfied:
  [interface org.jetbrains.kotlin.resolve.lazy.DeclarationScopeProvider]```). 

This does not make sense to me because I thought all the compiler modules were included in the compilerModules array. DeclarationScopeProvider lives in the frontend module, which is included in that list. The embedded(kotlinBuiltins()) does not work because kotlinBuiltins returns an Any, not a string.

Reply:
> resolving the LazyTopDownAnalyzer because of the DeclarationScopeProvider cannot be resolved
It might be because some part of intellij related dependencies is missing which was previously packed in the compiler, try adding intellij-core to classpath.
>The embedded(kotlinBuiltins()) does not work because kotlinBuiltins returns an Any, not a string.
There is an overload for `Any` how does it fail exactly?
https://github.com/JetBrains/kotlin/blob/master/buildSrc/src/main/kotlin/configurations.kt#L20

Reply:
I've successfully added the kotlinBuiltins dependency as well as the protobuf dependency but I still have the unresolved  DeclarationScopeProvider issue. I checked the jar for each of its dependencies and they are all there. I also have intellij dependencies on the classpath now, so that cannot be the issue. I'm wondering if there's a injection specification that is missing from the packages

Reply:
I confirmed that the container that is failing is LazyResolveWithJava, but since it's configured on creation I do not think it is an issue with a missing specification

Reply:
I ran a diff on the working jar versus the not working jar, so I'll work my way through these. Here is the list if anything jumps out:
```
Only in WORKS/META-INF: AL2.0
Only in WORKS/META-INF: Core.xml
Only in WORKS/META-INF: CoreImpl.xml
Only in WORKS/META-INF: JavaPsiPlugin.xml
Only in WORKS/META-INF: LGPL2.1
Only in WORKS/META-INF: LICENSE
Only in WORKS/META-INF: NOTICE
Only in WORKS/META-INF: intellij.platform.core.kotlin_module
Only in WORKS/META-INF: jb
Only in WORKS/META-INF: kotlinx-collections-immutable.kotlin_module
Only in WORKS/META-INF/maven: com.google.guava
Only in WORKS/META-INF/maven: io.javaslang
Only in WORKS/META-INF/maven: log4j
Only in WORKS/META-INF/maven: one.util
Only in WORKS/META-INF/maven: org.fusesource.hawtjni
Only in WORKS/META-INF/maven: org.fusesource.jansi
Only in WORKS/META-INF/maven: org.jetbrains.intellij.deps.fastutil
Only in WORKS/META-INF/maven: org.jline
Only in WORKS/META-INF: native
Only in WORKS/META-INF/services: javax.annotation.processing.Processor
Only in WORKS/META-INF/services: org.jline.terminal.spi.JansiSupport
Only in WORKS/META-INF/services: org.jline.terminal.spi.JnaSupport
Only in WORKS/com/google: common
Only in WORKS/com/google: thirdparty
Only in WORKS/com/intellij: AbstractBundle.class
Only in WORKS/com/intellij: BundleBase.class
Only in WORKS/com/intellij: BundleUtil.class
Only in WORKS/com/intellij: Patches.class
Only in WORKS/com/intellij: ReviseWhenPortedToJDK.class
Only in WORKS/com/intellij: UtilBundle.class
Only in WORKS/com/intellij: codeInsight
Only in WORKS/com/intellij: codeInspection
Only in WORKS/com/intellij: codeWithMe
Only in WORKS/com/intellij: concurrency
Only in WORKS/com/intellij: core
Only in WORKS/com/intellij: diagnostic
Only in WORKS/com/intellij: execution
Only in WORKS/com/intellij: extapi
Only in WORKS/com/intellij: formatting
Only in WORKS/com/intellij: icons
Only in WORKS/com/intellij: ide
Only in WORKS/com/intellij: indentation
Only in WORKS/com/intellij: injected
Only in WORKS/com/intellij: jna
Only in WORKS/com/intellij: lang
Only in WORKS/com/intellij: lexer
Only in WORKS/com/intellij: mock
Only in WORKS/com/intellij: model
Only in WORKS/com/intellij: navigation
Only in WORKS/com/intellij: notebook
Only in WORKS/com/intellij: openapi
Only in WORKS/com/intellij: patterns
Only in WORKS/com/intellij: pom
Only in WORKS/com/intellij: psi
Only in WORKS/com/intellij: refactoring
Only in WORKS/com/intellij: reference
Only in WORKS/com/intellij: security
Only in WORKS/com/intellij: serialization
Only in WORKS/com/intellij: serviceContainer
Only in WORKS/com/intellij/testFramework: BinaryLightVirtualFile$1.class
Only in WORKS/com/intellij/testFramework: BinaryLightVirtualFile.class
Only in WORKS/com/intellij/testFramework: LightVirtualFile$1.class
Only in WORKS/com/intellij/testFramework: LightVirtualFile.class
Only in WORKS/com/intellij/testFramework: LightVirtualFileBase$1.class
Only in WORKS/com/intellij/testFramework: LightVirtualFileBase$MyVirtualFileSystem.class
Only in WORKS/com/intellij/testFramework: LightVirtualFileBase.class
Only in WORKS/com/intellij/testFramework: ReadOnlyLightVirtualFile.class
Only in WORKS/com/intellij/testFramework: TestModeFlagListener.class
Only in WORKS/com/intellij/testFramework: TestModeFlags.class
Only in WORKS/com/intellij: testIntegration
Only in WORKS/com/intellij: ui
Only in WORKS/com/intellij: usageView
Only in WORKS/com/intellij: util
Only in WORKS/com/intellij: xml
Only in WORKS: it
Only in WORKS: javaslang
Only in WORKS: javax
Binary files WORKS/kotlin-compiler-internal-test-framework-1.5.255-SNAPSHOT.jar and NOTWORKS/kotlin-compiler-internal-test-framework-1.5.255-SNAPSHOT.jar differ
Only in WORKS: kotlinx
Only in WORKS: misc
Only in WORKS: net
Only in WORKS/org: apache
Only in WORKS/org: fusesource
Only in WORKS/org: jdom
Only in WORKS/org/jetbrains/annotations: SystemDependent.class
Only in WORKS/org/jetbrains/annotations: SystemIndependent.class
Only in WORKS/org/jetbrains: concurrency
Only in WORKS/org/jetbrains: ide
Only in WORKS/org/jetbrains: org
Only in WORKS/org: picocontainer
```
The intellij core and intellij platform have already been added to the plugin test project so I don't believe those are the issue

Reply:
Ok I fixed it by adding some of the missing dependencies to the plugin (annotations and javax inject)

Reply:
Ok it works! I was able to find the missing pieces. updating soon